### PR TITLE
[IMP] mail: Add a filter to tracking

### PR DESCRIPTION
**Pull request by reth-odoo**
## Usecase

It is sometimes desirable to completely remove a field from the automated tracking message
when setting a custom message that already covers the change being made.

## Solution

Add an argument when setting a custom message
to (conditionally) filter out the tracking fields from the end message.

task-2974500

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_move_line.py [hunk 1]
```diff
@@ -1558,10 +1558,11 @@ def _search_panel_domain_image(self, field_name, domain, set_count=False, limit=
     # TRACKING METHODS
     # -------------------------------------------------------------------------
 
-    def _mail_track(self, tracked_fields, initial):
-        changes, tracking_value_ids = super()._mail_track(tracked_fields, initial)
-        if len(changes) > len(tracking_value_ids):
-            for i, changed_field in enumerate(changes):
+    def _mail_track(self, tracked_fields, initial, filtered_fields=None):
+        changes, tracking_value_ids, filtered_out_fields = super()._mail_track(tracked_fields, initial, filtered_fields=filtered_fields)
+        displayed_changes = changes-filtered_out_fields
```
Comment (thjo-odoo) [inline addons/account/models/account_move_line.py:1563]:
```suggestion
        displayed_changes = changes - filtered_out_fields
```

File: addons/mail/models/models.py [hunk 2]
```diff
@@ -22,45 +23,59 @@ def _valid_field_parameter(self, field, name):
...
+        filtered_out_fields = set()
+
         # generate tracked_values data structure: {'col_name': {col_info, new_value, old_value}}
         for col_name, col_info in tracked_fields.items():
             if col_name not in initial:
                 continue
             initial_value = initial[col_name]
             new_value = self[col_name]
 
+
```
Comment (thjo-odoo) [inline addons/mail/models/models.py:61]: Unnecessary change

File: addons/project/models/project.py [hunk 3]
```diff
@@ -2265,11 +2265,12 @@ def _message_auto_subscribe_followers(self, updated_values, default_subtype_ids)
                 pass
         return new_followers
 
-    def _mail_track(self, tracked_fields, initial_values):
-        changes, tracking_value_ids = super()._mail_track(tracked_fields, initial_values)
+    def _mail_track(self, tracked_fields, initial_values, filtered_fields=None):
+        changes, tracking_value_ids, filtered_out_fields = super()._mail_track(tracked_fields, initial_values, filtered_fields=filtered_fields)
+        displayed_changes = changes-filtered_out_fields
```
Comment (thjo-odoo) [inline addons/project/models/project.py:2270]: Same as before

**Review by thjo-odoo — COMMENTED** [review]
Hello @thiryr,

Same as for the enterprise PR: small changes but nothing critical.
Otherwise LGTM!

File: addons/mail/models/mail_message.py [hunk 4]
```diff
@@ -884,11 +884,13 @@ def _message_format(self, fnames, format_reply=True):
                 'id': message_sudo.author_guest_id.id,
                 'name': message_sudo.author_guest_id.name,
             } if message_sudo.author_guest_id else [('clear',)]
+            thread_record_sudo = False
             if message_sudo.model and message_sudo.res_id:
                 record_sudo = self.env[message_sudo.model].browse(message_sudo.res_id).sudo()
                 record_name = record_sudo.with_prefetch(thread_ids_by_model_name[message_sudo.model]).display_name
                 # if not a thread, display_name is the default subject
-                default_subject = record_sudo._message_compute_subject() if hasattr(record_sudo, '_message_compute_subject') else record_name
+                thread_record_sudo = record_sudo if self.env['ir.model']._get(message_sudo.model).is_mail_thread else False
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_message.py:892]: Isn't that change adding queries ?

File: addons/mail/models/mail_message.py [hunk 5]
```diff
@@ -903,14 +905,15 @@ def _message_format(self, fnames, format_reply=True):
                 'partners': [{'id': partner.id, 'name': partner.name} for partner in reactions.partner_id],
             } for content, reactions in reactions_per_content.items()]
             allowed_tracking_ids = message_sudo.tracking_value_ids.filtered(lambda tracking: not tracking.field_groups or self.env.is_superuser() or self.user_has_groups(tracking.field_groups))
+            displayed_tracking_ids = thread_record_sudo._track_filter_displayed_tracking_values(allowed_tracking_ids) if thread_record_sudo else allowed_tracking_ids
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_message.py:908]: Which makes me wonder: how would that happen to have a message and tracking values without model / res_id ? As anyway tracking values should be linked to fields hence to models, maybe we should skip those and avoid any issue / leak with wrong messages ?

File: addons/mail/models/mail_message.py [hunk 6]
```diff
@@ -1176,12 +1179,14 @@ def _cleanup_side_records(self):
...
     def _filter_empty(self):
         """ Return subset of "void" messages """
+        IRModel = self.env['ir.model']
         return self.filtered(
             lambda msg:
                 (not msg.body or tools.is_html_empty(msg.body)) and
                 (not msg.subtype_id or not msg.subtype_id.description) and
                 not msg.attachment_ids and
-                not msg.tracking_value_ids
+                (not IRModel._get(msg.model).is_mail_thread or
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_message.py:1188]: Not sure we should change the heuristic here ... somehow it has tracking values so it is not "empty", is it ? It is used mainly to know if we have to "reset" message, which is rather a technical test.

File: addons/mail/models/mail_thread.py [hunk 7]
```diff
@@ -520,6 +520,10 @@ def _track_discard(self):
         for id_ in self.ids:
             initial_values[id_] = None
 
+    def _track_filter_displayed_tracking_values(self, tracking_values):
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:523]: ``_track_filter_for_display`` maybe ?

File: addons/mail/models/mail_thread.py [hunk 8]
```diff
@@ -562,6 +566,15 @@ def _track_set_log_message(self, message):
         for id_ in self.ids:
             body_values[id_] = message
 
+    def _track_get_message(self, tracked_fields):
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:569]: ``_track_get_default_log_message`` ? otherwise it is a bit hard to distinguish _track_set_log_message and _track_get_message maybe ?

File: addons/mail/models/mail_thread.py [hunk 9]
```diff
@@ -622,19 +635,21 @@ def _message_track(self, fields_iter, initial_values_dict):
                      for col_name in changes)
             )
             author_id = authors[record.id].id if record.id in authors else None
+            # _set_log_message takes priority over _mail_track_get_message even if it's an empty string
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:638]: don't forget to update method name in comment accordingly :D

File: addons/test_mail/models/test_mail_models.py [hunk 10]
```diff
@@ -198,6 +207,18 @@ def _notify_get_recipients_groups(self, message, model_description, msg_vals=Non
 
         return groups
 
+    def _track_filter_displayed_tracking_values(self, tracking_values):
```
Comment (tde-banana-odoo) [inline addons/test_mail/models/test_mail_models.py:210]: This model is used in performance tests (via 'mail.test.ticket.mc') ... would try to avoid modifying it and keep the same behavior if possible, it allows to compare counters across versions as behavior is always the same ^^ . Do you need those changes ?

File: addons/test_mail/models/test_mail_models.py [hunk 11]
```diff
@@ -112,7 +112,16 @@ class MailTestStandard(models.Model):
...
+    track_always_hide_tracking = fields.Boolean()
+    track_message = fields.Char(default='')
 
+    def _track_filter_displayed_tracking_values(self, tracking_values):
+        if self.track_always_hide_tracking:
+            return self.env['mail.tracking.value']
+        return super()._track_filter_displayed_tracking_values(tracking_values)
+
+    def _track_get_message(self, changes):
+        return self.track_message
```
Comment (tde-banana-odoo) [inline addons/test_mail/models/test_mail_models.py:124]: Maybe we could have something based on changes (like displaying changes), just to have a more dynamic test ?

File: addons/test_mail/tests/test_message_track.py [hunk 12]
```diff
@@ -35,6 +35,84 @@ def test_message_track_author(self):
 
         self.assertEqual(self.record.message_ids.author_id, self.partner_admin)
 
+    def test_message_track_default_message(self):
+        def _track_get_message(self, field_names):
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_track.py:39]: Can't we use `mail.test.track`model and avoid patching `mail.test.ticket` ? Just for simplicity of tests ?

File: addons/test_mail/tests/test_message_track.py [hunk 13]
```diff
@@ -35,6 +35,84 @@ def test_message_track_author(self):
...
+            self.record._track_set_log_message('Hi')
+            self.record.customer_id = False
+            self.flush_tracking()
+
+        messages = self.record.message_ids - messages
+        self.assertEqual(len(messages), 1)
+        self.assertEqual(messages.body, '<p>Hi</p>', '_track_set_log_message should take priority over default message')
+
+    def test_track_formatting(self):
+        def _track_filter_displayed_tracking_values(self, tracking_values):
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_track.py:63]: Same here, can't we paly with 'mail.test.track' whose method seems available for testing both behaviors (with and without filtering) ?

**Review by tde-banana-odoo — COMMENTED** [review]
Thanks for updated-updated-updated-updated version :D ! Looking good and quite simple indeed :) . Approach is good for me, added some comments while passing through.

File: addons/mail/models/mail_message.py (hunk 4 above)
Comment (thiryr) [reply addons/mail/models/mail_message.py]: I don't really like relying on hasattr but yes technically it does add a query :/

Will revert and add another hasattr later to be technically complete.

File: addons/mail/models/mail_message.py (hunk 5 above)
Comment (thiryr) [reply addons/mail/models/mail_message.py]: The idea here was that there isn't a *thread* record, but there should still be a record. It's just not useful if it's not a thread.

We could add an explicit check but in theory if you're able to read a message and the tracking values are not restricted

File: addons/mail/models/mail_message.py (hunk 6 above)
Comment (thiryr) [reply addons/mail/models/mail_message.py]: :thinking: yeah it doesn't make sense in a technical context, reverted

File: addons/test_mail/models/test_mail_models.py (hunk 10 above)
Comment (thiryr) [reply addons/test_mail/models/test_mail_models.py]: Think these were mainly 'needed' for the _is_empty test, which we're removing anyway :)

File: addons/test_mail/models/test_mail_models.py [hunk 14]
```diff
@@ -143,7 +143,18 @@ class MailTestStandard(models.Model):
     user_id = fields.Many2one('res.users', 'Responsible', tracking=True)
     container_id = fields.Many2one('mail.test.container', tracking=True)
     company_id = fields.Many2one('res.company')
-
+    track_hidden_fields = fields.Char()  # comma-separated list
```
Comment (tde-banana-odoo) [inline addons/test_mail/models/test_mail_models.py:146]: Maybe we should be more test-oriented, like
```suggestion
    track_fields_tofilter = fields.Char()  # comma-separated list
```

File: addons/test_mail/models/test_mail_models.py [hunk 15]
```diff
@@ -143,7 +143,18 @@ class MailTestStandard(models.Model):
...
+
+    def _track_filter_for_display(self, tracking_values):
+        values = super()._track_filter_for_display(tracking_values)
+        hidden_fields = set(self.track_hidden_fields.split(',') if self.track_hidden_fields else '')
+        return values.filtered(lambda val: val.field_id.name not in hidden_fields)
+
+    def _track_get_default_log_message(self, changes):
+        hidden_fields = set(self.track_hidden_fields.split(',') if self.track_hidden_fields else '')
+        if not all(change in hidden_fields for change in changes):
+            return f'There was a change on {self.name} for fields "{",".join(changes)}"'
```
Comment (tde-banana-odoo) [inline addons/test_mail/models/test_mail_models.py:156]: To avoid changes in test_mail_message.py, maybe
```suggestion
        if filtered_fields:
            return f'Filtered fields {",".join(changes)} have been updated on {self.name}, triggering custom log."
```

File: addons/test_mail/tests/test_mail_message.py [hunk 16]
```diff
@@ -45,7 +45,10 @@ def test_empty_message(self):
             'res_id': 0,
             'res_model': 'mail.compose.message',
         })
-        record = self.env['mail.test.track'].create({'name': 'EmptyTesting'})
+        record = self.env['mail.test.track'].create({
+            'name': 'EmptyTesting',
+            'track_hidden_fields': ','.join(self.env['mail.test.track']._fields)
+        })
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_message.py:51]: Continuing previous comment: would say that if track_hidden_fields is void, behavior is like before so that we don't impact tests.

File: addons/test_mail/tests/test_message_track.py [hunk 17]
```diff
@@ -22,7 +22,11 @@ def setUpClass(cls):
         ).with_context(cls._test_context).create({
             'name': 'Test',
         })
+        track_record = cls.env['mail.test.track'].with_user(cls.user_employee).with_context(cls._test_context).create({
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_track.py:25]: As it is used only 'test_message_track_default_message' we can probably create it in that test directly

File: addons/test_mail/tests/test_message_track.py [hunk 18]
```diff
@@ -39,6 +43,77 @@ def test_message_track_author(self):
 
         self.assertEqual(self.record.message_ids.author_id, self.partner_admin)
 
+    def test_message_track_default_message(self):
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_track.py:46]: Would use a '@users' to have ACLs activated :)  + don't forget docstring explaining we test '_track_get_default_log_message'

File: addons/test_mail/tests/test_message_track.py [hunk 19]
```diff
@@ -39,6 +43,77 @@ def test_message_track_author(self):
...
+        with self.mock_mail_gateway():
+            record._track_set_log_message('Hi')
+            record.user_id = False
+            self.flush_tracking()
+
+        messages = record.message_ids - messages
+        self.assertEqual(len(messages), 1)
+        self.assertEqual(messages.body, '<p>Hi</p>', '_track_set_log_message should take priority over default message')
+
+    def test_track_formatting(self):
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_track.py:66]: Formatting itself is already tests, maybe just target the filtering itself ?
```suggestion
    def test_track_filter_for_display(self):
```

+ docstring and @users decorator :)

File: addons/test_mail/tests/test_message_track.py [hunk 20]
```diff
@@ -39,6 +43,77 @@ def test_message_track_author(self):
...
+             'track_hidden_fields': 'user_id' if n == 0 else ''
+             } for n in range(2)])
+        self.flush_tracking()
+
+        records.write({'user_id': new_user.id})
+        self.flush_tracking()
+
+        for record in records:
+            self.assertEqual(len(record.message_ids), 2, 'Should be a creation message and a tracking message')
+            self.assertEqual(len(record.message_ids[0].tracking_value_ids), 1, 'Latest message should have a single tracking value')
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_track.py:85]: Maybe use 'assertTracking' ?

File: addons/test_mail/tests/test_message_track.py [hunk 21]
```diff
@@ -39,6 +43,77 @@ def test_message_track_author(self):
...
+        field_name = 'user_id'
+        field_type = 'many2one'
+        original_user = self.user_admin
+        new_user = self.user_employee
+
+        records = self.env['mail.test.track'].create([
+            {'name': f'Test{n}',
+             'user_id': original_user.id,
+             'track_hidden_fields': 'user_id' if n == 0 else ''
+             } for n in range(2)])
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_track.py:77]: Isn't it possible to create two records manually ^^ ? Seems it would be easier to understand what you want to do :)

File: addons/test_mail/tests/test_message_track.py [hunk 22]
```diff
@@ -39,6 +43,77 @@ def test_message_track_author(self):
...
+        self.flush_tracking()
+
+        records.write({'user_id': new_user.id})
+        self.flush_tracking()
+
+        for record in records:
+            self.assertEqual(len(record.message_ids), 2, 'Should be a creation message and a tracking message')
+            self.assertEqual(len(record.message_ids[0].tracking_value_ids), 1, 'Latest message should have a single tracking value')
+
+        # hidden
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_track.py:87]: Maybe
```suggestion
        # first record: tracking value should be hidden (see '_track_filter_for_display')
```

File: addons/test_mail/tests/test_message_track.py [hunk 23]
```diff
@@ -39,6 +43,77 @@ def test_message_track_author(self):
...
+        message_0 = records[0].message_ids[0]
+        formatted = message_0.message_format()[0]
+        self.assertEqual(formatted['trackingValues'], [], 'Hidden values should not be formatted')
+        mail_render = records[0]._notify_by_email_prepare_rendering_context(message_0, {})
+        self.assertEqual(mail_render['tracking_values'], [])
+
+        # visible
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_track.py:94]:
```suggestion
        # second record: all values displayed
```

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, some comments (mainly on tests) ! LGTM otherwise, and I think we can target 17.0 for this one :)

File: addons/test_mail/models/test_mail_models.py (hunk 15 above)
Comment (thiryr) [reply addons/test_mail/models/test_mail_models.py]: I feel like that's too specific to the one test. I'll just add a field to enable it, not like it costs anything, and I guess it's more test-oriented too :)

**Comment by tde-banana-odoo** [comment]
@robodoo r+
